www/docs/server/validators.md

```diff
@@ -0,0 +1,321 @@
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+// ---cut---
+
+export const appRouter = t.router({
+  hello: t.procedure
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const name = opts.input.name;
+      //    ^?
+      return {
+        greeting: `Hello ${opts.input.name}`,
+      };
+    }),
+});
+```
+
+`.input` can also be stacked to build more complex types, which is particularly useful when you want to utilise some common input to a collection of procedures in a [middleware](middlewares).
+
+```ts twoslash
+// @target: esnext
+import { TRPCError, initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+
+// ---cut---
+
+const procedureWhichRejectsOutOfTownFolk = t.procedure
+  .input(z.object({ townName: z.string() }))
+  .use((opts) => {
+    if (opts.input.townName !== 'Pucklechurch') {
+      throw new TRPCError({
+        code: 'FORBIDDEN',
+        message: "We don't take kindly to out-of-town folk",
+      });
+    }
+
+    return opts.next();
+  });
+
+export const appRouter = t.router({
+  hello: procedureWhichRejectsOutOfTownFolk
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const input = opts.input;
+      //    ^?
+      return {
+        greeting: `Hello ${input.name}, my friend`,
+      };
+    }),
+});
+```
+
+### Output Validators
+
+By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.
+
+Validators are used to infer the input and output types of your procedures.
+
+:::info
+If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
+:::
+
+## Using a function as a validator
+
+If you like, you can simply pass a function as a validator, though this isn't necessarily the best choice to use widely unless you have a specific need.
+
+In most cases we recommend you use a [Validator](#validator-integrations)
+
+```ts twoslash
+import { initTRPC } from '@trpc/server';
+
+export const t = initTRPC.create();
+
+export const appRouter = t.router({
+  hello: t.procedure
+    .input((value: any): { name: string } => {
+      if (value && typeof value.name === 'string') {
+        return { name: value.name };
+      }
+      throw new Error('Greeting not found');
+    })
+    .output((value: any) => {
+      if (value && typeof value.greeting === 'string') {
+        return { greeting: value.greeting };
+      }
+      throw new Error('Greeting not found');
+    })
+    .query(({ input }) => {
+      //      ^?
+      return {
+        greeting: `hello ${input.name}`,
+      };
+    }),
+});
+
+export type AppRouter = typeof appRouter;
+```
+
+## Validator integrations
```

Comment:
tabs would be nice to compact this down

Reply:
C&P from the old thread as I guess it got lost:

I had a look at this and felt that actually the way things are is quite nice. Having a lot of linkable sections in the sidebar should pay us back later in visibility and linkability. Plus the list if quite long and may not scale well to tabs as it continues to grow

Reply:
This is quite pleasant in the sidebar

<img width="301" alt="image" src="https://user-images.githubusercontent.com/8896153/234409312-f8e1dd81-d016-4a53-9631-64f88e45574a.png">